src/contracts/pods/EigenPod.sol

```diff
@@ -169,133 +140,114 @@ contract EigenPod is IEigenPod, Initializable, ReentrancyGuardUpgradeable, Eigen
+            beaconTimestamp != 0, 
+            "EigenPod.verifyCheckpointProofs: must have active checkpoint to perform checkpoint proof"
         );
 
-        // Verify passed-in beaconStateRoot against oracle-provided block root:
-        BeaconChainProofs.verifyStateRootAgainstLatestBlockRoot({
-            latestBlockRoot: eigenPodManager.getBlockRootAtTimestamp(oracleTimestamp),
-            beaconStateRoot: stateRootProof.beaconStateRoot,
-            stateRootProof: stateRootProof.proof
-        });
+        Checkpoint memory checkpoint = currentCheckpoint;
 
-        VerifiedWithdrawal memory withdrawalSummary;
-        for (uint256 i = 0; i < withdrawalFields.length; i++) {
-            VerifiedWithdrawal memory verifiedWithdrawal = _verifyAndProcessWithdrawal(
-                stateRootProof.beaconStateRoot,
-                withdrawalProofs[i],
-                validatorFieldsProofs[i],
-                validatorFields[i],
-                withdrawalFields[i]
-            );
+        // If we haven't already proven a state root for this checkpoint, verify
+        // `stateRootProof` against the block root and cache the result
+        if (checkpoint.beaconStateRoot == bytes32(0)) {
+            BeaconChainProofs.verifyStateRootAgainstLatestBlockRoot({
+                latestBlockRoot: checkpoint.beaconBlockRoot,
+                beaconStateRoot: stateRootProof.beaconStateRoot,
+                stateRootProof: stateRootProof.proof
+            });
 
-            withdrawalSummary.amountToSendGwei += verifiedWithdrawal.amountToSendGwei;
-            withdrawalSummary.sharesDeltaGwei += verifiedWithdrawal.sharesDeltaGwei;
+            checkpoint.beaconStateRoot = stateRootProof.beaconStateRoot;
         }
 
-        // If any withdrawals are eligible for immediate redemption, send to the pod owner via
-        // DelayedWithdrawalRouter
-        if (withdrawalSummary.amountToSendGwei != 0) {
-            _sendETH_AsDelayedWithdrawal(podOwner, withdrawalSummary.amountToSendGwei * GWEI_TO_WEI);
-        }
-        // If any withdrawals resulted in a change in the pod's shares, update the EigenPodManager
-        if (withdrawalSummary.sharesDeltaGwei != 0) {
-            eigenPodManager.recordBeaconChainETHBalanceUpdate(podOwner, withdrawalSummary.sharesDeltaGwei * int256(GWEI_TO_WEI));
+        // Process each checkpoint proof submitted
+        for (uint256 i = 0; i < proofs.length; i++) {
+
+            // Process a checkpoint proof for a validator. 
+            // - the validator MUST be in the ACTIVE state
+            // - the validator MUST NOT have already been proven for this checkpoint
+            //
+            // If the proof shows the validator has a balance of 0, they are marked `WITHDRAWN`.
+            // The assumption is that if this is the case, any withdrawn ETH was already in
+            // the pod when `startCheckpoint` was originally called.
```

Comment:
is this an OK assumption? seems like the ETH could ultimately be recovered/proven in a subsequent checkpoint?
mostly making sure there is no case where ETH could get stuck.

Reply:
Should be a good assumption! It's kinda the main thing we need to be true, so if there are problems with it, speak up!

I describe how the assumption works here: https://hackmd.io/@-HV50kYcRqOjl_7du8m1AA/SkJPfqBeC#What-Gets-Snapshotted